src/host/readDataCooked.cpp

```diff
@@ -436,633 +196,1027 @@ bool COOKED_READ_DATA::AtEol() const noexcept
+    _bufferCursor = _buffer.size();
+    _markAsDirty();
+}
+
+// If the viewport is large enough to fit a popup, this function prepares everything for
+// showing the given type. It handles computing the size of the popup, its position,
+// backs the affected area up and draws the border and initial contents.
+void COOKED_READ_DATA::_popupPush(const PopupKind kind)
+try
 {
-    size_t bytesToWrite = 0;
-    const auto hr = SizeTMult(cch, sizeof(wchar_t), &bytesToWrite);
-    if (FAILED(hr))
+    auto& textBuffer = _screenInfo.GetTextBuffer();
+    const auto viewport = _screenInfo.GetViewport();
+    const auto viewportOrigin = viewport.Origin();
+    const auto viewportSize = viewport.Dimensions();
+
+    til::size proposedSize;
+    switch (kind)
+    {
+    case PopupKind::CopyToChar:
+        proposedSize = { 26, 1 };
+        break;
+    case PopupKind::CopyFromChar:
+        proposedSize = { 28, 1 };
+        break;
+    case PopupKind::CommandNumber:
+        proposedSize = { 22 + CommandNumberMaxInputLength, 1 };
+        break;
+    case PopupKind::CommandList:
     {
-        return 0;
+        const auto& commands = _history->GetCommands();
+        const auto commandCount = _history->GetNumberOfCommands();
+
+        size_t maxStringLength = 0;
+        for (const auto& c : commands)
+        {
+            maxStringLength = std::max(maxStringLength, c.size());
+        }
+
+        // Account for the "123: " prefix each line gets.
+        maxStringLength += integerLog10(commandCount);
+        maxStringLength += 3;
+
+        // conhost used to draw the command list with a size of 40x10, but at some point it switched over to dynamically
+        // sizing it depending on the history count and width of the entries. Back then it was implemented with the
+        // assumption that the code unit count equals the column count, which I kept because I don't want the TextBuffer
+        // class to expose how wide characters are, any more than necessary. It makes implementing Unicode support
+        // much harder, because things like combining marks and work zones may cause TextBuffer to end up deciding
+        // a piece of text has a different size than what you thought it had when measuring it on its own.
+        proposedSize.width = gsl::narrow_cast<til::CoordType>(std::clamp<size_t>(maxStringLength, 40, til::CoordTypeMax));
+        proposedSize.height = std::clamp(commandCount, 10, 20);
+        break;
+    }
+    default:
+        assert(false);
+        return;
     }
 
-    memmove(_userBuffer, _backupLimit, bytesToWrite);
+    // Subtract 2 because we need to draw the border around our content. We must return early if we're
+    // unable to do so, otherwise the remaining code fails because the size would be zero/negative.
+    const til::size contentSize{
+        std::min(proposedSize.width, viewportSize.width - 2),
+        std::min(proposedSize.height, viewportSize.height - 2),
+    };
+    if (!contentSize)

```

Comment:
if only one dimension is negative, should we bail? (I don't think it does in the new version, dunno about old)

Reply:
Hm... Yeah I think we should bail. It's easier to write that way (we got the `operator!` for `til::size`) and the situation won't occur in practice IMO.